{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ðŸ”¥ðŸ”¥ TrajectoryBuilder is now running in BERTRAM_MODE ðŸ”¥ðŸ”¥\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.model_selection import RandomizedSearchCV\n",
    "\n",
    "from sign.CONST import AMOUNT_OF_KEYFRAMES\n",
    "\n",
    "from sign.training.load_data.dynamic_loader import DynamicLoader\n",
    "\n",
    "CSV_PATH = 'dyn_50_common_youtube_55p.csv'\n",
    "NUM_HANDS = 2\n",
    "\n",
    "dyna_lod = DynamicLoader(AMOUNT_OF_KEYFRAMES)\n",
    "training_data = dyna_lod.prepare_training_data(CSV_PATH)\n",
    "xs,ys,xs_test,ys_test = training_data.spliterino(random_state=1)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# LogisticRegression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(LogisticRegression(C=930, max_iter=10000, random_state=42),\n",
       " 6,\n",
       " {'C': 930},\n",
       " 0.5352519009026157)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "\n",
    "logistic_reg = LogisticRegression(random_state=42,max_iter=10000,C=90)\n",
    "parameters = {'C':range(1000,10000,200)}\n",
    "pre_clf = RandomizedSearchCV(logistic_reg,parameters,n_jobs=-1)\n",
    "pre_clf.fit(np.array(xs),ys)\n",
    "print(pre_clf.best_estimator_,pre_clf.best_index_,pre_clf.best_params_,pre_clf.best_score_)\n",
    "print(classification_report(ys_test,pre_clf.predict(np.array(xs_test))))\n",
    "\n",
    "if False:\n",
    "    from joblib import dump\n",
    "    dump(pre_clf, 'dynamic_lr.joblib')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# SVC / SVM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(SVC(C=400.0, gamma=0.4, random_state=42),\n",
       " 0,\n",
       " {'gamma': 0.4, 'C': 400.0},\n",
       " 0.5751985724095497)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.preprocessing import StandardScaler, PolynomialFeatures\n",
    "from sklearn.pipeline import make_pipeline\n",
    "\n",
    "svc = make_pipeline(StandardScaler(),SVC())\n",
    "\n",
    "parameters = {'C':[float(x) for x in range(0,2000,200)],\"gamma\":[x * 0.1 for x in range(0,51,1)]}\n",
    "pre_clf = RandomizedSearchCV(svc,parameters,n_jobs=-1)\n",
    "pre_clf.fit(np.array(xs),ys)\n",
    "print(pre_clf.best_estimator_,pre_clf.best_index_,pre_clf.best_params_,pre_clf.best_score_)\n",
    "print(classification_report(ys_test,pre_clf.predict(np.array(xs_test))))\n",
    "\n",
    "if False:\n",
    "    from joblib import dump\n",
    "    dump(model, 'dynamic_svm.joblib')\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Random Forest Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "random_forest = RandomForestClassifier()\n",
    "parameters = {\"max_features\":[\"sqrt\",\"log2\"],\"n_estimators\":range(100,1001,100)}\n",
    "pre_clf = RandomizedSearchCV(random_forest,parameters)\n",
    "pre_clf.fit(np.array(xs),ys)\n",
    "print(pre_clf.best_estimator_,pre_clf.best_index_,pre_clf.best_params_,pre_clf.best_score_)\n",
    "print(classification_report(ys_test,pre_clf.predict(np.array(xs_test))))\n",
    "\n",
    "if False:\n",
    "    from joblib import dump\n",
    "    dump(pre_clf, 'dynamic_rf.joblib')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Voting Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_search.py:318: UserWarning: The total space of parameters 2 is smaller than n_iter=10. Running 2 iterations. For exhaustive searches, use GridSearchCV.\n",
      "  warnings.warn(\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:993: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/svm/_base.py\", line 823, in _check_proba\n",
      "    raise AttributeError(\n",
      "AttributeError: predict_proba is not available when probability=False\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/pipeline.py\", line 53, in check\n",
      "    getattr(self._final_estimator, attr)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'SVC' has no attribute 'predict_proba'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py\", line 982, in _score\n",
      "    scores = scorer(estimator, X_test, y_test, **score_params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/metrics/_scorer.py\", line 415, in __call__\n",
      "    return estimator.score(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/base.py\", line 764, in score\n",
      "    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)\n",
      "                             ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 383, in predict\n",
      "    maj = np.argmax(self.predict_proba(X), axis=1)\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 424, in predict_proba\n",
      "    self._collect_probas(X), axis=0, weights=self._weights_not_none\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in _collect_probas\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in <listcomp>\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                       ^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'Pipeline' has no attribute 'predict_proba'\n",
      "\n",
      "  warnings.warn(\n",
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:993: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/svm/_base.py\", line 823, in _check_proba\n",
      "    raise AttributeError(\n",
      "AttributeError: predict_proba is not available when probability=False\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/pipeline.py\", line 53, in check\n",
      "    getattr(self._final_estimator, attr)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'SVC' has no attribute 'predict_proba'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py\", line 982, in _score\n",
      "    scores = scorer(estimator, X_test, y_test, **score_params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/metrics/_scorer.py\", line 415, in __call__\n",
      "    return estimator.score(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/base.py\", line 764, in score\n",
      "    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)\n",
      "                             ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 383, in predict\n",
      "    maj = np.argmax(self.predict_proba(X), axis=1)\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 424, in predict_proba\n",
      "    self._collect_probas(X), axis=0, weights=self._weights_not_none\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in _collect_probas\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in <listcomp>\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                       ^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'Pipeline' has no attribute 'predict_proba'\n",
      "\n",
      "  warnings.warn(\n",
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:993: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/svm/_base.py\", line 823, in _check_proba\n",
      "    raise AttributeError(\n",
      "AttributeError: predict_proba is not available when probability=False\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/pipeline.py\", line 53, in check\n",
      "    getattr(self._final_estimator, attr)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'SVC' has no attribute 'predict_proba'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py\", line 982, in _score\n",
      "    scores = scorer(estimator, X_test, y_test, **score_params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/metrics/_scorer.py\", line 415, in __call__\n",
      "    return estimator.score(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/base.py\", line 764, in score\n",
      "    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)\n",
      "                             ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 383, in predict\n",
      "    maj = np.argmax(self.predict_proba(X), axis=1)\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 424, in predict_proba\n",
      "    self._collect_probas(X), axis=0, weights=self._weights_not_none\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in _collect_probas\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in <listcomp>\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                       ^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'Pipeline' has no attribute 'predict_proba'\n",
      "\n",
      "  warnings.warn(\n",
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:993: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/svm/_base.py\", line 823, in _check_proba\n",
      "    raise AttributeError(\n",
      "AttributeError: predict_proba is not available when probability=False\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/pipeline.py\", line 53, in check\n",
      "    getattr(self._final_estimator, attr)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'SVC' has no attribute 'predict_proba'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py\", line 982, in _score\n",
      "    scores = scorer(estimator, X_test, y_test, **score_params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/metrics/_scorer.py\", line 415, in __call__\n",
      "    return estimator.score(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/base.py\", line 764, in score\n",
      "    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)\n",
      "                             ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 383, in predict\n",
      "    maj = np.argmax(self.predict_proba(X), axis=1)\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 424, in predict_proba\n",
      "    self._collect_probas(X), axis=0, weights=self._weights_not_none\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in _collect_probas\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in <listcomp>\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                       ^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'Pipeline' has no attribute 'predict_proba'\n",
      "\n",
      "  warnings.warn(\n",
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py:993: UserWarning: Scoring failed. The score on this train-test partition for these parameters will be set to nan. Details: \n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/svm/_base.py\", line 823, in _check_proba\n",
      "    raise AttributeError(\n",
      "AttributeError: predict_proba is not available when probability=False\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 29, in _check\n",
      "    check_result = self.check(obj)\n",
      "                   ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/pipeline.py\", line 53, in check\n",
      "    getattr(self._final_estimator, attr)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'SVC' has no attribute 'predict_proba'\n",
      "\n",
      "The above exception was the direct cause of the following exception:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_validation.py\", line 982, in _score\n",
      "    scores = scorer(estimator, X_test, y_test, **score_params)\n",
      "             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/metrics/_scorer.py\", line 415, in __call__\n",
      "    return estimator.score(*args, **kwargs)\n",
      "           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/base.py\", line 764, in score\n",
      "    return accuracy_score(y, self.predict(X), sample_weight=sample_weight)\n",
      "                             ^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 383, in predict\n",
      "    maj = np.argmax(self.predict_proba(X), axis=1)\n",
      "                    ^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 424, in predict_proba\n",
      "    self._collect_probas(X), axis=0, weights=self._weights_not_none\n",
      "    ^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in _collect_probas\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/ensemble/_voting.py\", line 399, in <listcomp>\n",
      "    return np.asarray([clf.predict_proba(X) for clf in self.estimators_])\n",
      "                       ^^^^^^^^^^^^^^^^^\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 40, in __get__\n",
      "    self._check(obj, owner=owner)\n",
      "  File \"/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/utils/_available_if.py\", line 31, in _check\n",
      "    raise AttributeError(attr_err_msg) from e\n",
      "AttributeError: This 'Pipeline' has no attribute 'predict_proba'\n",
      "\n",
      "  warnings.warn(\n",
      "/mnt/repositories/Bsc-Sign-Language-Recon/backend/.venv/lib/python3.11/site-packages/sklearn/model_selection/_search.py:1051: UserWarning: One or more of the test scores are non-finite: [0.59135921        nan]\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "from sklearn.ensemble import VotingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.preprocessing import StandardScaler, PolynomialFeatures\n",
    "from sklearn.svm import SVC\n",
    "from sklearn.pipeline import make_pipeline\n",
    "\n",
    "lr = LogisticRegression(C=930.0,max_iter=10_000)\n",
    "rf = RandomForestClassifier(random_state=42,n_estimators=500,max_features=\"sqrt\")\n",
    "svc = make_pipeline(StandardScaler(),SVC(random_state=42,probability=True,gamma=0.4,C=400.0)) # remember probability\n",
    "voting_clf = VotingClassifier( estimators=[ \n",
    "                                           ('lr', lr), \n",
    "                                           ('rf', rf), \n",
    "                                           ('svc', svc)\n",
    "]) \n",
    "parameters = {\"voting\":[\"hard\",\"soft\"]}\n",
    "pre_clf = RandomizedSearchCV(voting_clf,parameters,n_jobs=2)\n",
    "pre_clf.fit(np.array(xs),ys)\n",
    "print(\"VotingClassifier\")\n",
    "print(pre_clf.best_estimator_,pre_clf.best_index_,pre_clf.best_params_,pre_clf.best_score_)\n",
    "print(classification_report(ys_test,pre_clf.predict(xs_test)))\n",
    "\n",
    "if False:\n",
    "    from joblib import dump\n",
    "    dump(pre_clf, 'dynamic_vc.joblib')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Bagging Classifier"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sklearn.ensemble import BaggingClassifier\n",
    "from sklearn.linear_model import LogisticRegression\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "from sklearn.svm import SVC\n",
    "lr = LogisticRegression(C=930.0,max_iter=10_000)\n",
    "rf = RandomForestClassifier(random_state=42,n_estimators=500,max_features=\"sqrt\")\n",
    "svc = make_pipeline(StandardScaler(),SVC(random_state=42,probability=True,gamma=0.4,C=400.0))\n",
    "\n",
    "bagging_clf = BaggingClassifier(estimator=[ \n",
    "                                           ('lr', lr), \n",
    "                                           ('rf', rf), \n",
    "                                           ('svc', svc)\n",
    "])\n",
    "\n",
    "bagging_clf.fit(xs, ys)\n",
    "parameters = {\"voting\":[\"hard\",\"soft\"]}\n",
    "pre_clf = RandomizedSearchCV(bagging_clf,parameters,n_jobs=-1)\n",
    "pre_clf.fit(np.array(xs),ys)\n",
    "print(\"VotingClassifier\")\n",
    "print(pre_clf.best_estimator_,pre_clf.best_index_,pre_clf.best_params_,pre_clf.best_score_)\n",
    "print(classification_report(ys_test,pre_clf.predict(xs_test)))\n",
    "\n",
    "if False:\n",
    "    from joblib import dump\n",
    "    dump(pre_clf, 'dynamic_bc.joblib')\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
